File: packages/sdk/src/resources/pricing.ts
import type { TheatricalHTTPClient } from '../http/client';
import type {
  PriceCalculation,
  TicketType,
  TicketTypeFilter,
  ApplyCouponsInput,
  CouponApplicationResult,
} from '../types/pricing';

/**
 * Pricing resource — ticket types, price calculations, tax handling, and coupon application.
 *
 * Handles the complexity of Vista's pricing model: matinee vs peak pricing,
 * format surcharges (IMAX, 4DX), loyalty tier discounts, and coupon stacking.
 */
export class PricingResource {
  constructor(private readonly http: TheatricalHTTPClient) {}

  /**
   * List available ticket types for a session.
   *
   * @param sessionId - The Vista session ID.
   * @param filter - Optional filter params (category, availableOnly).
   */
  async ticketTypes(sessionId: string, filter?: Omit<TicketTypeFilter, 'sessionId'>): Promise<TicketType[]> {
    return this.http.get<TicketType[]>(`/ocapi/v1/sessions/${sessionId}/ticket-types`, {
      params: { ...filter },
    });
  }

  /**
   * Calculate the price for a ticket type and quantity.
   *
   * Returns an itemised {@link PriceBreakdown} including base price, discounts,
   * surcharges, and tax. The breakdown reflects the caller's Vista site tax
   * configuration (inclusive vs exclusive).
   *
   * @param sessionId - The Vista session ID.
   * @param ticketTypeId - Ticket type to price.
   * @param quantity - Number of tickets (default 1).
   */
  async calculate(sessionId: string, ticketTypeId: string, quantity = 1): Promise<PriceCalculation> {
    return this.http.get<PriceCalculation>(`/ocapi/v1/pricing/calculate`, {
      params: { sessionId, ticketTypeId, quantity },
    });
  }

  /**
   * Apply coupon codes and resolve loyalty-tier discounts.
   *
   * Validates each coupon code, stacks applicable discounts, and returns
   * an updated {@link PriceBreakdown}. Invalid or expired codes are listed
   * in the `rejected` array rather than throwing.
   *
   * @param input - Session, ticket type, quantity, coupon codes, and optional member ID.
   */
  async applyCoupons(input: ApplyCouponsInput): Promise<CouponApplicationResult> {
    return this.http.post<CouponApplicationResult>(`/ocapi/v1/pricing/apply-coupons`, { body: input });
  }
}

File: packages/sdk-py/theatrical/resources/pricing.py
"""Pricing resource — ticket types, price calculations, and coupon application."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

from pydantic import TypeAdapter

from theatrical.types.pricing import (
    ApplyCouponsInput,
    CouponApplicationResult,
    PriceCalculation,
    TicketType,
    TicketTypeFilter,
)

if TYPE_CHECKING:
    from theatrical.http.client import TheatricalHttpProtocol

_ticket_type_list_adapter = TypeAdapter(list[TicketType])
_price_calc_adapter = TypeAdapter(PriceCalculation)
_coupon_result_adapter = TypeAdapter(CouponApplicationResult)


class PricingResource:
    def __init__(self, http: TheatricalHttpProtocol) -> None:
        self._http = http

    async def ticket_types(
        self, session_id: str, filters: Optional[TicketTypeFilter] = None
    ) -> list[TicketType]:
        params: dict[str, str] = {}
        if filters:
            d = filters.model_dump(exclude_none=True, exclude={"session_id"})
            for key, value in d.items():
                params[key] = str(value) if not isinstance(value, str) else value
        raw = await self._http.get(
            f"/ocapi/v1/sessions/{session_id}/ticket-types", params=params or None
        )
        return _ticket_type_list_adapter.validate_python(raw)

    async def calculate(
        self,
        session_id: str,
        ticket_type_id: str,
        quantity: int = 1,
    ) -> PriceCalculation:
        raw = await self._http.get(
            "/ocapi/v1/pricing/calculate",
            params={
                "sessionId": session_id,
                "ticketTypeId": ticket_type_id,
                "quantity": str(quantity),
            },
        )
        return _price_calc_adapter.validate_python(raw)

    async def apply_coupons(self, input: ApplyCouponsInput) -> CouponApplicationResult:
        raw = await self._http.post(
            "/ocapi/v1/pricing/apply-coupons",
            body=input.model_dump(exclude_none=True),
        )
        return _coupon_result_adapter.validate_python(raw)
